minetest/minetest
# [NOSQUASH] Smart ptr stuff from sound refactor PR, and add make_irr

Changes taken from #12764, to make that PR simpler.

Unique and irr ptrs are used for GUIEngine's members, and for Game's sound manager.

The global dummySoundManager is removed, together with client/sound.cpp.

Note: In GUIEngine, a ptr to the local `MenuMusicFetcher soundfetcher` variable was passed to the sound manager, I think this was a stack UaF. Fixed (therefore this PR is a bugfix). (Or it was no UaF because the sound manager was never accessed after `soundfetcher` goes out of scope. But it's still super fishy.)

Also adds a `make_irr` helper (like `make_unique`).

## To do

This PR is a Ready for Review.

## How to test

* Build.
* Check that the game still launches and works.
* Look through code. There should be no functional changes.
